wf/clean.py

```diff
@@ -0,0 +1,172 @@
+    # compared to either rows or columns.
+
+    deviations = int(deviations)
+    
+    str_length_r = singlecell['row'].unique().tolist()
+    all_indexes_r = {}
+    for i in str_length_r: 
+      pre_list = np.where(singlecell['row'] == i)
+      indexes = pre_list[0].tolist()
+      pre_sort = singlecell.iloc[indexes]['passed_filters'].tolist()
+      pre_sort.sort()
+      all_indexes_r[i] = statistics.median(pre_sort)
+    
+    
+    str_length_c = singlecell['col'].unique().tolist()
+    all_indexes_c = {}
+    for i in str_length_c: 
+      pre_list = np.where(singlecell['col'] == i)
+      indexes = pre_list[0].tolist()
+      pre_sort = singlecell.iloc[indexes]['passed_filters'].tolist()
+      pre_sort.sort()
+      all_indexes_c[i] = statistics.median(pre_sort)
+    
+    apply_medians_c = singlecell.loc[:,'col'].apply(lambda x: all_indexes_c[x])
+
+    row_mean = statistics.mean(list(all_indexes_r.values()))
+    row_std = statistics.stdev(list(all_indexes_r.values()))
+    col_mean = statistics.mean(list(all_indexes_c.values()))
+    col_std = statistics.stdev(list(all_indexes_c.values()))
+    # identify limit more than x standard deviations above mean
+    rows_limit = row_mean + deviations * row_std
+    cols_limit = col_mean + deviations * col_std
+
+    # create table with only diagonal tixels from singlecell table
+    diag_sc = singlecell[singlecell['row'] == singlecell['col']]
+    diag_mean = diag_sc['passed_filters'].mean()
+
+    # create 'adjust' column with reads to downsample
+    if diag_mean > rows_limit:
+        diag_sc['adjust'] = np.ceil(diag_sc['passed_filters'] * (row_mean / diag_mean))
+    elif diag_mean > cols_limit:
+        diag_sc['adjust'] = np.ceil(diag_sc['passed_filters'] * (col_mean / diag_mean))
+    else:
+        diag_sc['adjust'] = diag_sc['passed_filters']
+
+    return diag_sc[['barcode', 'adjust']]
+
+def combine_tables(singlecell, deviations=1):
+    row_singlecell = singlecell.copy()
+    col_singlecell = singlecell.copy()
+    dia_singlecell = singlecell.copy()
+    row_reductions = get_reductions(row_singlecell, 'row', deviations)
+    col_reductions = get_reductions(col_singlecell, 'col', deviations)
+    diag_reductions = get_diag_reductions(dia_singlecell, deviations)
+
+    # concat rows and columns
+    # If a tixel occurs twice, take the average value
+    combined_table = averageDuplicates([row_reductions.values.tolist(), col_reductions.values.tolist(), diag_reductions.values.tolist()])
+    
+    return combined_table
+
+def clean_fragments(fragments_path, r_table):
+    # Reduce high tixels by randomly downsampling fragments.tsv according to
+    # reduction table
+    skip_headers = [i+1 for i in range(50)]
+    print("Loading fragments.tsv")
+    # fragments = gzip.open(fragments_path)
+    fragments = pd.read_csv(fragments_path, sep='\t', header=skip_headers, skiprows=skip_headers)
+    fragments.columns = ['V1', 'V2', 'V3', 'barcode', 'V4']
```

Comment:
Looks like using header=skip_headers actually moves the first 50 rows into the header; try, 
```suggestion
fragments = pd.read_csv(fragments_path, sep='\t', header=None, skiprows=list(range(51)))
fragments.columns = ['V1', 'V2', 'V3', 'barcode', 'V4']
```